xhydro/modelling/_ravenpy_models.py

```diff
@@ -0,0 +1,199 @@
+
+
+class RavenpyModel(HydrologicalModel):
+    r"""Implement the RavenPy model class to build and run ravenpy models.
+
+    Parameters
+    ----------
+    model_name : str
+        The name of the ravenpy model to run. Can be one of
+        ["blended", "gr4jcn", "hbvec", "hmets", "hypr", "mohyse", "sacsma"].
+    parameters : np.ndarray
+        The model parameters for simulation or calibration.
+    drainage_area : float
+        The watershed drainage area, in km².
+    elevation : float
+        The elevation of the watershed, in meters.
+    latitude : float
+        The latitude of the watershed centroid.
+    longitude : float
+        The longitude of the watershed centroid.
+    start_date : dt.datetime
+        The first date of the simulation.
+    end_date : dt.datetime
+        The last date of the simulation.
+    qobs_path : Union[str, os.PathLike]
+        The path to the dataset containing the observed streamflow.
+    alt_names_flow : dict
+        A dictionary that allows users to change the names of flow variables of their dataset to cf-compliant names.
+    meteo_file : Union[str, os.PathLike]
+        The path to the file containing the observed meteorological data.
+    data_type : dict
+        The dictionary necessary to tell raven which variables are being fed such that it can adjust it's processes
+        internally.
+    alt_names_meteo : dict
+        A dictionary that allows users to change the names of meteo variables of their dataset to cf-compliant names.
+    meteo_station_properties : dict
+        The properties of the weather stations providing the meteorological data. Used to adjust weather according to
+        differences between station and catchment elevations (adiabatic gradients, etc.).
+    rain_snow_fraction : str
+        The method used by raven to split total precipitation into rain and snow.
+    evaporation : str
+        The evapotranspiration function used by raven.
+    \**kwargs : dict
+        Dictionary of other parameters to feed to raven according to special cases and that are allowed by the raven
+        documentation.
+    """
+
+    def __init__(
+        self,
+        model_name: str,
+        parameters: np.ndarray,
+        drainage_area: Union[str, os.PathLike],
+        elevation: str,
+        latitude,
+        longitude,
+        start_date,
+        end_date,
+        qobs_path,
+        alt_names_flow,
+        meteo_file,
+        data_type,
+        alt_names_meteo,
+        meteo_station_properties,
+        rain_snow_fraction="RAINSNOW_DINGMAN",
+        evaporation="PET_PRIESTLEY_TAYLOR",
+        **kwargs,
+    ):
+        # Create HRU object for ravenpy based on catchment properties
+        self.model_simulations = None
+        self.qsim = None
+        hru = dict(
+            area=drainage_area,
+            elevation=elevation,
+            latitude=latitude,
+            longitude=longitude,
+            hru_type="land",
+        )
```

Comment:
On va probablement aussi vouloir qu'un utilisateur puisse pointer vers un fichier de HRU déjà construit? Je pense ici au travail pour l'Atlas, par exemple.

Reply:
Hmmm c'est un bon point. En fait, ce qu'on voudrait, c'est qu'un usager puisse prendre ses fichiers .rvX au complet et rouler dans xhydro comme n'importe quel modèle. On a des outils pour ça aussi. Laisse-moi voir ce que je peux faire.

Reply:
Bon, à bien y penser, je ne suis pas certain que ça soit idéal d'avoir les modèles custom dans xhydro (du moins pour Ravenpy). Ravenpy a déjà un mécanisme permettant de "loader" un modèle à partir de sa configuration déjà construite et de le faire tourner. Le code pour le faire est très simple:

```
from ravenpy import OutputReader
from ravenpy.ravenpy import run

# Get the .rvX files paths
config = [
    get_file(f"raven-gr4j-cemaneige/raven-gr4j-salmon.{ext}")
    for ext in ["rvt", "rvc", "rvi", "rvh", "rvp"]
]

configdir = config[0].parent

outputs_path = run(modelname="raven-gr4j-salmon", configdir=configdir, overwrite=True)
outputs = OutputReader(run_name=None, path=outputs_path)
```
Cependant, on ne peut pas calibrer ces modèles car ils peuvent avoir toutes sortes de paramètres custom. Si on fait un modèle spécifique pour le projet (ex. ravenpy HBVEC distribué) alors je crois que c'est mieux d'avoir une fonction spécifique pour celui-là.

Le HRU tel qu'il est indiqué ici est la manière de faire pour les modèles qu'on a "en stock" dans ravenpy actuellement alors je ne le changerais pas.

Qu'en penses-tu?

Reply:
Est-ce que l'idée serait que l'utilisateur puisse faire tourner le modèle 1) à partir de ses fichiers .rvx ou 2) d'une shapefile des HRUs avec un NetCDF des inputs météo ? Dans le cas 1) on n'aurait même pas besoin de créer la configuration de l'émulateur ("# Create the emulator configuration") si je ne me trompe pas ? Si on préfère aller vers l'option 2), je suggérerais d'aller vers quelque chose d'un peu plus flexible pour créer le .rvh de HRUs et de remplacer cette partie du code (lignes 95-101) avec :

HRUs_gdf = gpd.read_file(hru_file.shp)
rvh_config = BasinMakerExtractor(HRUs_gdf).extract()

Puis ensuite lorsque l'on crée l'émulateur (ligne 104) on peut fournir au dictionnaire :

# The hydrological network
sub_basins=rvh_config['sub_basins'],
sub_basin_group=rvh_config['sub_basin_group'], # optionnel ?
reservoirs=rvh_config['reservoirs'], # optionnel
channel_profile=rvh_config['channel_profile'],
hrus=rvh_config['hrus'],

Qu'en pensez vous ?